pan3d/dataset_builder.py

```diff
@@ -0,0 +1,284 @@
+import json
+import os
+import pyvista
+import xarray
+from pvxarray.vtk_source import PyVistaXarraySource
+from pyvista.trame.ui import plotter_ui
+
+from trame.decorators import TrameApp, change
+from trame.ui.vuetify import SinglePageWithDrawerLayout  # TODO: upgrade to vuetify 3
+from trame.app import get_server
+
+from pan3d.ui import AxisSelection, MainDrawer, RenderArea, Toolbar
+from pan3d.utils import initial_state, run_singleton_task
+
+
+@TrameApp()
+class DatasetBuilder:
+    def __init__(self, server=None, dataset_path=None, state=None):
+        if server is None:
+            server = get_server()
+
+        if isinstance(server, str):
+            server = get_server(server)
+
+        self.server = server
+        self.state.update(initial_state)
+        self.algorithm = PyVistaXarraySource()
+        self.plotter = pyvista.Plotter(off_screen=True, notebook=False)
+        self.plotter.set_background("lightgrey")
+        self.dataset = None
+        self.dataset_path = None
+        self.mesh = None
+        self.actor = None
+
+        self.ctrl.get_plotter = lambda: self.plotter
+        self.ctrl.reset = self.reset
+
+        if dataset_path:
+            self.set_dataset_path(dataset_path=dataset_path)
+        if state:
+            self.state.update(state)
+
+        # Fix version of vue
+        server.client_type = "vue2"  # TODO: upgrade to vue3
+
+        # Build UI
+        self.layout = SinglePageWithDrawerLayout(self.server)
+        self.layout.title.set_text("Pan3D Viewer")
+        self.layout.footer.hide()
+        with self.layout:
+            with self.layout.toolbar:
+                self.layout.toolbar.dense = True
+                self.layout.toolbar.align = "center"
+                Toolbar(reset=self.ctrl.reset)
+            with self.layout.drawer:
+                MainDrawer()
+            with self.layout.content:
+                plot_view = plotter_ui(
+                    self.ctrl.get_plotter(),
+                    interactive_ratio=1,
+                )
+                self.ctrl.view_update = plot_view.update
+                self.ctrl.reset_camera = plot_view.reset_camera
+                RenderArea(plot_view=plot_view)
+            AxisSelection()
+
+    # -----------------------------------------------------
+    # Properties
+    # -----------------------------------------------------
+
+    @property
+    def state(self):
+        return self.server.state
+
+    @property
+    def ctrl(self):
+        return self.server.controller
+
+    @property
+    def viewer(self):
+        return self.layout
```

Comment:
This could encapsulate line 50-65 with some lazy loading logic.

Reply:
```python

@property
    def viewer(self):
        if self.layout is None:
              # l50-65
        return self.layout
```

Reply:
Ok, changed in [a13307f](https://github.com/Kitware/pan3d/pull/28/commits/a13307f4fb78d9a4a6d0b29dda9d3cac963df7f1)